libfwupd/fwupd-common.c

```diff
@@ -397,46 +397,55 @@ fwupd_build_user_agent_system(void)
 gchar *
 fwupd_build_machine_id(const gchar *salt, GError **error)
 {
-	const gchar *fn = NULL;
+	const gchar *machine_id;
+	gsize bufsz = 0;
 	g_autofree gchar *buf = NULL;
-	g_auto(GStrv) fns = g_new0(gchar *, 6);
 	g_autoptr(GChecksum) csum = NULL;
-	gsize sz = 0;
 
 	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
 
-	/* one of these has to exist */
-	fns[0] = g_build_filename(FWUPD_SYSCONFDIR, "machine-id", NULL);
-	fns[1] = g_build_filename(FWUPD_LOCALSTATEDIR, "lib", "dbus", "machine-id", NULL);
-	fns[2] = g_strdup("/etc/machine-id");
-	fns[3] = g_strdup("/var/lib/dbus/machine-id");
-	fns[4] = g_strdup("/var/db/dbus/machine-id");
-	for (guint i = 0; fns[i] != NULL; i++) {
-		if (g_file_test(fns[i], G_FILE_TEST_EXISTS)) {
-			fn = fns[i];
-			break;
+	/* in test mode */
+	machine_id = g_getenv("FWUPD_MACHINE_ID");
+	if (machine_id != NULL) {
+		buf = g_strdup(machine_id);
+		bufsz = strlen(buf);
+	} else {
+		const gchar *fn = NULL;
+		g_auto(GStrv) fns = g_new0(gchar *, 6);
+
+		/* one of these has to exist */
+		fns[0] = g_build_filename(FWUPD_SYSCONFDIR, "machine-id", NULL);
+		fns[1] = g_build_filename(FWUPD_LOCALSTATEDIR, "lib", "dbus", "machine-id", NULL);
```

Comment:
This feels like it should be using get path and then tests should write machine ID to that path instead of env variable.

Reply:
> should be using get path

This is libfwupd, not libfwupdplugin tho.

Reply:
Maybe we need a similar helper in a static function then?

Reply:
I'm not even sure I *like* the design we have in libfwupdplugin -- if feels like something that should be stored in the context rather than all those env variables.

Reply:
>  maybe we should be using umockdev more for testing

Yes, although this is pretty cumbersome -- look at the thunderbolt plugin for example. I'm wondering if we want something like I did with the USB emulation; i.e. emulate stuff in sysfs from a JSON file too.

Reply:
Take a look how it's done in power profiles daemon. Python code that you can make helpers and stuff.

Reply:
Ack; lets merge this first. I'm a bit disappointed the coverage numbers went down, as all of these commits were to address stuff we were not testing.